src/command_buffer.rs

```diff
@@ -0,0 +1,220 @@
+        }
+        let begin = self.entities.last().map_or(0, |x| x.end);
+        let end = begin + len;
+        self.entities.push(EntityIndex {
+            entity: ent,
+            begin,
+            end,
+        });
+    }
+
+    /// Spawn every entity recorded with their components
+    ///
+    /// # Example
+    /// ```
+    /// # use hecs::*;
+    /// let mut world = World::new();
+    /// let a = world.reserve_entity();
+    /// let mut recorder = CommandBuffer::new();
+    /// recorder.spawn_at(a, (false,0.0));
+    /// recorder.run_on(&mut world);
+    /// assert!(world.contains(a));
+    /// ```
+    pub fn run_on(&mut self, world: &mut World) {
+        let mut mark = self.entities.len() - 1;
+
+        for index in self.entities.iter() {
+            self.info[index.begin..index.end].sort_unstable_by_key(|z| z.ty_info);
+        }
+
+        for _ in 0..self.entities.len() {
+            let (ent, comps) = self.build(mark);
+            world.spawn_at(ent, comps);
+            if mark != 0 {
+                mark -= 1;
+            }
+        }
+        self.clear();
+    }
+
+    fn build(&mut self, mark: usize) -> (Entity, ReadyBuffer<'_>) {
+        self.ids.extend(
+            self.info[self.entities[mark].begin..self.entities[mark].end]
+                .iter()
+                .map(|x| x.ty_info.id()),
+        );
+        let entity = self.entities[mark].entity;
+        (entity, ReadyBuffer { buffer: self, mark })
+    }
+
+    /// Drop previously `recorded` entities and their components
+    ///
+    /// Recorder is cleared implicitly when entities are spawned, so usually this doesn't need to
+    /// be called
+    pub fn clear(&mut self) {
+        self.ids.clear();
+        self.entities.clear();
+        self.cursor = 0;
+        unsafe {
+            for info in self.info.drain(..) {
+                info.ty_info.drop(self.storage.as_ptr().add(info.offset));
+            }
+        }
+    }
+}
+
+unsafe impl Send for CommandBuffer {}
+unsafe impl Sync for CommandBuffer {}
+
+impl Drop for CommandBuffer {
+    fn drop(&mut self) {
+        self.clear();
+        if self.layout.size() != 0 {
+            unsafe {
+                dealloc(self.storage.as_ptr(), self.layout);
+            }
+        }
+    }
+}
+
+impl Default for CommandBuffer {
+    /// Create an empty buffer
+    fn default() -> Self {
+        Self {
+            entities: Vec::new(),
+            storage: NonNull::dangling(),
+            layout: Layout::from_size_align(0, 8).unwrap(),
+            cursor: 0,
+            info: Vec::new(),
+            ids: Vec::new(),
+        }
+    }
+}
+
+/// The output of an '[CommandBuffer]` suitable for passing to
+/// [`World::spawn_into`](crate::World::spawn_into)
+struct ReadyBuffer<'a> {
+    buffer: &'a mut CommandBuffer,
+    mark: usize,
+}
+unsafe impl DynamicBundle for ReadyBuffer<'_> {
+    fn with_ids<T>(&self, f: impl FnOnce(&[TypeId]) -> T) -> T {
+        f(&self.buffer.ids)
```

Comment:
This passes the *entire* list of IDs, which may have been `extend`ed multiple times. Is that intended?

Given the issues here, I think this module should have a unit test that verifies correct behavior for these methods.

Reply:
good catch this caused a very sneaky bug, I fixed it and added a test for it, for other methods tests I provided before should do the work.